### PR TITLE
Optimize the nested for-loop (#26)

### PR DESCRIPTION
**Pull request by taizan-hokuto**
## 関連Issue
#26 （YunzheZJU さん）
#28

2つ以上のライブ配信がある場合にも対応しました。（ほとんどYunzheZJUさんが示してくださったコードのおかげです。ありがとうございます）
うまくページを取得できない場合は時間をずらしながらリトライするようにしてみました。

### CODE REVIEW

**Comment by hidori-11** [comment]
コード拝見させていただきました～
Session張りっぱなしにしたんですね
毎回張り直すよりはいいかもですね

追記
張り直すって表現まちがってるな・・・？　保持って言ったほうがいいか

**Comment by taizan-hokuto** [comment]
レスありがとうございます。手のひら返しですが、接続が継続すればオーケーと思い、保持するようにしてみました。

File: mikochiku_alarm.py [hunk 1]
```diff
@@ -135,45 +150,34 @@ def alarm_sound(self):
...
+            video_id_set = set(video_ids)
             return video_id_set
+        except requests.exceptions.ConnectTimeout as e:
+            print('タイムアウトしました。')
+            # sessionを張り直して継続する。
+            self.session = self.reset_session()
+            return set()
+        except requests.exceptions.RequestException as e:
+            if response.status_code == 404:
+                print(f'{search_ch_id} は、存在しないチャンネルです。')
```
Comment (YunzheZJU) [inline mikochiku_alarm.py:171]: Is there any value supposed to be returned here?
I think `self.session.close()` @ line 176 will be executed if no `return` here, which means a 404 Error will cause the program to abort. Is this expected?

Comment (taizan-hokuto) [reply mikochiku_alarm.py:171]: >Is there any value supposed to be returned here?

None. 
I expected the program would stop if a 404 error occurred.
 
However, I have realized that we are able to change or delete the channel ID in the GUI. Therefore, it is better to notice a warning of invalid channel ID and continue program execution than to abort.

The current implementation does not display the console (pyinstaller --no-console), so I think that it is user-friendly if there is a GUI that pops out errors etc.

Comment (YunzheZJU) [reply mikochiku_alarm.py:171]: Thank you for your explanation!

**Comment by pusaitou** [comment]
以前プルリクをくれたapocelipesさんが #2 でprint("")をつかったデバッグの代わりにpyqtのqDebug("")を用いていました
これがpyinstaller --no-consoleの状態でも使えるデバッグ関数なのかな？と思っています

**Comment by pusaitou** [comment]
他のプログラムでrequestsを使っているとエラー処理の部分でたまにこんなのが出ることがあります

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connection.py", line 159, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw)
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\util\connection.py", line 57, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "C:\ProgramData\Anaconda3\lib\socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 11001] getaddrinfo failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connectionpool.py", line 343, in _make_request
    self._validate_conn(conn)
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connectionpool.py", line 839, in _validate_conn
    conn.connect()
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connection.py", line 301, in connect
    conn = self._new_conn()
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connection.py", line 168, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
urllib3.exceptions.NewConnectionError: <urllib3.connection.VerifiedHTTPSConnection object at 0x0000022CBEC897F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\requests\adapters.py", line 449, in send
    timeout=timeout
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "C:\ProgramData\Anaconda3\lib\site-packages\urllib3\util\retry.py", line 398, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='www.youtube.com', port=443): Max retries exceeded with url: /channel/UC-hM6YJuNYVAmUWxeIr9FeA (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x0000022CBEC897F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "miko_alert.py", line 153, in <module>
    html = requests.get("https://www.youtube.com/channel/" + search_ch_id)
  File "C:\ProgramData\Anaconda3\lib\site-packages\requests\api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\requests\api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\requests\sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\requests\sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\requests\adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='www.youtube.com', port=443): Max retries exceeded with url: /channel/UC-hM6YJuNYVAmUWxeIr9FeA (Caused by NewConnectionError('<urllib3.connection.VerifiedHTTPSConnection object at 0x0000022CBEC897F0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))


この最初のsocket.gaierrorのエラーがでるとこのプログラムだと異常終了しないですか？

**Comment by pusaitou** [comment]
このプログラムはヘッダを設定しないでrequests.get()しているのでyoutubeから得られるhtmlが変化しています
それも影響しているかもしれません

**Comment by pusaitou** [comment]
エラー処理ってぜんぜんやらないので勝手がわからないのをお許しください

**Comment by taizan-hokuto** [comment]
現行はrequestsのエラー発生時リトライするようにしていますが、gai.socketerrorだとすり抜ける可能性があるため、
まずtestにてgai.socketerrorを再現する等して、リトライされるか確認します。
そのうえで、gai.socketerror発生時もリトライされない場合はリトライされるように改修。

あとは、実運用上でgai.socketerrorを含めた接続エラーが頻繁に発生するかどうかです。

実運用上でgai.socketerror等接続エラーが発生する主原因が、youtubeからの接続拒否と思われる場合（たとえば〇〇時間以上起動しっぱなしにしていると必ずgai.socketerrorが発生する等）、対処方法を別途考える必要があります。

（コネクション間隔（40秒）を長くする／リトライ間隔を長くする／ヘッダを変えて再接続を試みる／proxyを使う→最後はそこまでやる必要性はあるか検討の余地はありますが）

あとは自分が設定した現行のリトライ間隔(接続タイムアウト20秒、読込タイムアウト10秒）が短くないか、リトライ回数が少なくないか（3回）も検討の余地あり。

**Comment by YunzheZJU** [comment]
Close #26

**Comment by pusaitou** [comment]
なるほど、そのようにしてエラーをつぶしていけばいいんですね！gai.socketerrorが発生するのは結構ランダムですが今のところ二日に一回とか発生することもあれば数時間で数回発生することもあります
ヘッダをランダムに変更して毎回接続するのは今回の件の対策に限らずやってみたいやつですね

少しここで提案があります
今回作っているのはアラームで常駐させるソフトウェアなのでいつの間にかソフトウェアが落ちていて鳴らなかったということは避けたいんです
未知のエラーはすべてreturnして次のrequestsしてくる情報に懸けるみたいな実装にしてみたいんですが危険性はないでしょうか？

**Comment by taizan-hokuto** [comment]
>未知のエラーはすべてreturnして次のrequestsしてくる情報に懸けるみたいな実装にしてみたいんですが危険性はないでしょうか？

例外処理で補足できるエラーである限り、エラーが発生しても落とさず監視を継続する形で実装することはできると思います。

「危険性」ですが、どのような危険性を考えられているのでしょうか。
BAN的なものに関する危険なのか、プログラムの継続動作上の危険なのか。

前者については、運用してみないとわかりません。
別issueで、1秒間隔アクセスを続けて半日IP BANみたいなこともあったそうなので（ #29 ）。

後者については、たとえばメモリリークでずっと動かしてるとハングする、とか？

**Comment by pusaitou** [comment]
エラーが発生しても落とさず監視を継続する形がいいです！
危険性についてあまり具体的なものが見えてこなかったので提示してもらえて助かります
前者の危険性について前者のBAN的なものですが、IPBANを食らってもブラウザからアクセスすると人間かどうかを判断するCAPTCHAが表れてそれをクリアすればyoutubeを継続してみることができました。なのでみこ畜アラームだけが使えなくなる程度の問題です。たまにCAPTCHAすら現れないこともありましたがすぐにCAPTCHAが表れるページが表れたので実害は比較的少ないかと。みこ畜アラームを多数起動して自分でIPBANになったら使用者の責任ですからあまり気にしなくていいと思ってます。
後者のプログラミングの継続動作上の危険はメモリーリークに関してはあんまり起きなさそうなので対応しなくてもいいと思ってます。
よってすごく暇なときにIPBANされたことをユーザーに伝えるシステムが実装されればいいかなと思ってるくらいです。それよりも実装しなければいけないところがあるのでそれを優先したいと思います。

**Pull request merged**